src/python/tests/appengine/libs/request_cache_test.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""request_cache tests."""
+
+from builtins import object
+import unittest
+
+from libs import request_cache
+from tests.test_libs import helpers as test_helpers
+
+
+class CacheClass(object):
+  """Test cache class."""
+
+  def __init__(self):
+    self.called = []
+
+  @request_cache.wrap(3)
+  def foo(self, a):
+    self.called.append((a,))
+    return a + 1
+
+
+class CacheClass2(object):
+  """Test cache class."""
+
+  def __init__(self):
+    self.called = []
+
+  @request_cache.wrap(3)
+  def foo(self, a):
+    self.called.append((a,))
+    return a + 2
+
+
+class CacheTest(unittest.TestCase):
+  """Cache tests."""
+
+  def setUp(self):
+    test_helpers.patch(self, [
+        'webapp2.get_request',
+    ])
+
+    self.mock.get_request.return_value = object()
+
+  def test_basic(self):
+    """Basic tests."""
+    c = CacheClass()
+    self.assertEqual(2, c.foo(1))
+    self.assertEqual(2, c.foo(1))
+    self.assertEqual(3, c.foo(2))
+
+    self.assertListEqual([
+        (1,),
+        (2,),
+    ], c.called)
+
+  def test_no_request(self):
+    """Test no request available."""
+    self.mock.get_request.return_value = None
+    c = CacheClass()
+    self.assertEqual(2, c.foo(1))
+    self.assertEqual(2, c.foo(1))
+
```

Comment:
do we want to self.assertEqual(3, c.foo(2)) with different arg as well

Reply:
Done!